include/itkBlockMatchingBayesianRegularizationDisplacementCalculator.hxx

```diff
@@ -344,19 +360,28 @@ BayesianRegularizationDisplacementCalculator< TMetricImage, TDisplacementImage >
     m_PriorPr = this->m_MetricImageImage;
     this->m_MetricImageImage = tempMetricImageImagePtr;
 
-    this->ApplyThreadFunctor( m_CopyPriorToPosteriorThreadFunctor );
+    this->m_MultiThreader->template ParallelizeImageRegion<ImageDimension>(
+      this->m_DisplacementImage->GetRequestedRegion(),
+      [this]( const RegionType & outputRegion )
+          {
+          this->ThreadedCopyPriorToPosterior( outputRegion );
+          },
+      nullptr);
 
+    const ThreadIdType workUnits = this->m_MultiThreader->GetNumberOfWorkUnits();
     for( fit = faceList.begin(); fit != faceList.end(); ++fit )
       {
-      ImpartLikelihoodThreadStruct str;
-      str.self = this;
-      this->m_DisplacementImage->SetRequestedRegion( *fit );
-      this->m_Threader->SetSingleMethod(
-        this->ImpartLikelihoodThreaderCallback, &str );
-      this->m_Threader->SingleMethodExecute();
+      // TODO: FIX-ME: hangs if this is not set
+      this->m_MultiThreader->SetNumberOfWorkUnits( 2 );
```

Comment:
@dzenanz can you please look into this? I observe a hang with the `itk::PoolMultiThreader`. I suspect that the region is smaller than the number of work units requested, but the threading backends should be able to handle this condition gracefully.

Reply:
It doesn't hang on my Windows computer even if I remove `this->m_MultiThreader->SetNumberOfWorkUnits( 2 );` I will test on Linux too.

Reply:
On Linux I had [this](http://review.source.kitware.com/#/c/23576/) branch built, and with it there is no hang, whether using Pool or TBB. I didn't test with master on Linux.

Reply:
Likely related -- this was testing on a 12-core system.

However, I tried with the `PoolMultiThreader` refactoring patch, and it passed! :fireworks:

Reply:
Excellent! Once PoolMultiThreader refactoring gets merged, we can remove this workaround and merge this and/or #81.

Reply:
PoolMultiThreader refactoring has been merged, so this is good to go I suppose.

Reply:
Don't forget to remove this.